### PR TITLE
HIP Introduction

### PR DESCRIPTION
**Pull request by neon60**

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -1,27 +1,20 @@
...
-For HIP supported AMD GPUs on multiple operating systems, see:
-
-* [Linux system requirements](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/reference/system-requirements.html#supported-gpus)
-* [Microsoft Windows system requirements](https://rocm.docs.amd.com/projects/install-on-windows/en/latest/reference/system-requirements.html#windows-supported-gpus)
-
-The CUDA enabled NVIDIA GPUs are supported by HIP. For more information, see [GPU Compute Capability](https://developer.nvidia.com/cuda-gpus).
-
-On the AMD ROCm platform, HIP provides header files and runtime library built on top of HIP-Clang compiler in the repository [Common Language Runtime (CLR)](./understand/amd_clr), which contains source codes for AMD's compute languages runtimes as follows,
-
-On non-AMD platforms, like NVIDIA, HIP provides header files required to support non-AMD specific back-end implementation in the repository ['hipother'](https://github.com/ROCm/hipother), which translates from the HIP runtime APIs to CUDA runtime APIs.
```
Comment (neon60) [inline docs/index.md:16]: Yes, I also do not like to extend this section with random links. Where should we move this?

File: docs/index.md [hunk 2]
```diff
@@ -1,27 +1,20 @@
...
-
-## Overview
+The HIP documentation is organized as follows:
 
 ::::{grid} 1 1 2 2
 :gutter: 3
 
 :::{grid-item-card} Install
 
+* [Linux system requirements](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/reference/system-requirements.html#supported-gpus)
```
Comment (neon60) [inline docs/index.md]: @randyh62 The original link point us to the supported GPUs and you also remove the Windows supported GPUs.  Also removed the CUDA enabled GPUs list: "The CUDA enabled NVIDIA GPUs are supported by HIP. For more information, see [GPU Compute Capability](https://developer.nvidia.com/cuda-gpus)."

We can add linux and windows requirements links to install, but we need to also have a link about the supported GPUs. We can add that to Install or Reference. Maybe link the requirements at install and add the supported GPUs to reference?

File: docs/index.md [hunk 3]
```diff
@@ -1,27 +1,20 @@
 # HIP documentation
 
 The Heterogeneous-computing Interface for Portability (HIP) API is a C++ runtime
-API and kernel language that lets developers create portable applications for AMD
-and NVIDIA GPUs from single source code.
+API and kernel language that lets developers create portable applications running
+in heterogeneous systems, using CPUs and AMD GPUs and NVIDIA GPUs from a single source code.
```
Comment (randyh62) [inline docs/index.md]: in heterogeneous systems using CPUs, and AMD GPUs or NVIDIA GPUs, from a single source code.

File: docs/understand/introduction_to_hip.rst [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+For the AMD ROCm platform, HIP provides headers and a runtime library built on top of HIP-Clang compiler in the repository `Common Language Runtime (CLR) <./amd_clr.html>`_.  The HIP runtime implements HIP streams, events, and memory APIs, and is a object library that is linked with the application.  The source code for all headers and the library implementation is available on GitHub. HIP developers on ROCm can use :doc:`ROCgdb <rocgdb:index>` for debugging and :doc:`ROCProfiler <rocprofiler:index>` for profiling.
```
Comment (randyh62) [inline docs/understand/introduction_to_hip.rst]: and is an object library that is linked with the application.

File: docs/index.md (hunk 1 above)
Comment (randyh62) [reply docs/index.md:16]: This content has been relocated to the Introduction, or the cards on the Index below this.

File: docs/understand/introduction_to_hip.rst [hunk 5]
```diff
@@ -10,19 +10,19 @@ Introduction to HIP
 
 The Heterogeneous-computing Interface for Portability (HIP) is a C++ runtime API and kernel language that lets you create portable applications for AMD and NVIDIA GPUs from a single source code. 
 
-* HIP is a thin API with very little performance impact over coding directly in NVIDIA CUDA or AMD ROCm.
+* HIP is a thin API with little or no performance impact over coding directly in NVIDIA CUDA or AMD ROCm.
```
Comment (MKKnorr) [inline docs/understand/introduction_to_hip.rst]: This solved the problem with the overhead over "programming in AMD ROCm", but still doesn't solve the underlying question of: what is programming in AMD ROCm? We should think of a better way of wording this, or straight up say it the way it is: for CUDA it is a wrapper, but for ROCm it's the "native" way of programming for AMD GPUs

File: docs/understand/introduction_to_hip.rst [hunk 6]
```diff
@@ -10,19 +10,19 @@ Introduction to HIP
...
+HIP supports the ability to build and run on either AMD GPUs or NVIDIA GPUs. GPU Programmers familiar with NVIDIA CUDA or OpenCL will find the HIP API familiar and easy to use. Developers no longer need to choose between AMD or NVIDIA GPUs. You can quickly port your application to run on the available hardware while maintaining a single codebase. The HIPify tools, based on the clang front-end and Perl language, can convert CUDA API calls into the corresponding HIP calls. However, HIP is not intended to be a drop-in replacement for CUDA, and developers should expect to do some manual coding and performance tuning work to port existing projects as described in `HIP Porting Guide <../how-to/hip_porting_guide.html>`_.
```
Comment (MKKnorr) [inline docs/understand/introduction_to_hip.rst]: Missed the "API" and there are two trailing empty spaces at the end of the line :)
```suggestion
HIP supports the ability to build and run on either AMD GPUs or NVIDIA GPUs. GPU Programmers familiar with NVIDIA CUDA or OpenCL will find the HIP API familiar and easy to use. Developers no longer need to choose between AMD or NVIDIA GPUs. You can quickly port your application to run on the available hardware while maintaining a single codebase. The HIPify tools, based on the clang front-end and Perl language, can convert CUDA API calls into the corresponding HIP API calls. However, HIP is not intended to be a drop-in replacement for CUDA, and developers should expect to do some manual coding and performance tuning work to port existing projects as described in `HIP Porting Guide <../how-to/hip_porting_guide.html>`_.
```

**Review by randyh62 — APPROVED** [review]

File: docs/understand/programming_model.rst [hunk 7]
```diff
@@ -78,7 +126,7 @@ mostly known from exploiting relevant instruction sets on CPUs (for example SSE/
 
 A HIP device compiler maps SIMT code written in HIP C++ to an inherently SIMD
 architecture (like GPUs). This is done by scalarizing the entire kernel and issuing the scalar
-instructions of multiple kernel instances to each of the SIMD engine lanes, rather
+instructions of multiple kernel instances (or threads) to each of the SIMD engine lanes, rather
```
Comment (MKKnorr) [inline docs/understand/programming_model.rst]:
```suggestion
instructions of multiple kernel instances (called threads) to each of the SIMD engine lanes, rather
```

File: docs/understand/programming_model.rst [hunk 8]
```diff
@@ -111,46 +160,214 @@ typically look the following:
...
+=====================
 
-* The C++ abstract machine assumes a unified memory address space, meaning that
-  one can always access any given address in memory (assuming the absence of
-  data races). HIP however introduces several memory namespaces, an address
-  from one means nothing in another. Moreover, not all address spaces are
-  accessible from all contexts.
+The SIMT nature of HIP is captured by the ability to execute user-provided
+device programs, expressed as single-source C/C++ functions or sources compiled
+online/offline to binaries in bulk.
```
Comment (MKKnorr) [inline docs/understand/programming_model.rst]:
```suggestion
online/offline to binaries, in bulk.
```

File: docs/understand/programming_model.rst [hunk 9]
```diff
@@ -111,46 +160,214 @@ typically look the following:
...
 
-  If one were to look at :ref:`cdna2_gcd` and inside the :ref:`cdna3_cu`,
-  every Compute Unit has an instance of storage backing the namespace
-  ``__shared__``. Even if the host were to have access to these regions of
-  memory, the performance benefits of the segmented memory subsystem are
-  supported by the inability of asynchronous access from the host.
+Multiple instances of the device program (or kernel) are called threads and may
```
Comment (MKKnorr) [inline docs/understand/programming_model.rst]: This has already been mentioned earlier in the SIMT section

File: docs/understand/introduction_to_hip.rst [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+* HIP is a thin API with little or no performance impact over coding directly in NVIDIA CUDA or AMD ROCm.
+* HIP enables coding in a single-source C++ programming language including features such as templates, C++11 lambdas, classes, namespaces, and more.
+* The :doc:`HIPify <hipify:index>` tools convert source from CUDA to HIP.
+* Developers can specialize for the platform (CUDA or AMD) to tune for performance or handle tricky cases.
+
+HIP includes the runtime API, kernel language, compilers (``clang``, ``hipcc``), code profilers (``rocprof``, ``omnitrace``), debugging tools (``rocgdb``), and libraries to create heterogeneous applications running on both CPUs and GPUs. HIP provides marshalling libraries like :doc:`hipFFT <hipfft:index>` or :doc:`hipBLAS <hipblas:index>` that act as a thin programming layer over either NVIDIA CUDA or AMD ROCm to enable support for either language as a back-end. These libraries offer pointer-based memory interfaces and are easily integrated into your applications.
```
Comment (MKKnorr) [inline docs/understand/introduction_to_hip.rst]: This still has to be separated between the CUDA and ROCm backend

File: docs/understand/programming_model.rst (hunk 7 above)
Comment (neon60) [reply docs/understand/programming_model.rst]: Fixed on the other PR.

File: docs/understand/programming_model.rst (hunk 8 above)
Comment (neon60) [reply docs/understand/programming_model.rst]: Fixed on the other PR.

File: docs/understand/programming_model.rst (hunk 9 above)
Comment (neon60) [reply docs/understand/programming_model.rst]: @MKKnorr I fixed at the other PR.

File: docs/what_is_hip.rst [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+.. figure:: data/what_is_hip/hip.svg
+    :alt: HIP in an application.
+    :align: center
+
+* HIP is a thin API with little or no performance impact over coding directly
+  in NVIDIA CUDA or AMD :doc:`ROCm <rocm:what_is_rocm>`.
+* HIP enables coding in a single-source C++ programming language including
+  features such as templates, C++11 lambdas, classes, namespaces, and more.
+* Developers can specialize for the platform (CUDA or AMD) to tune for
```
Comment (neon60) [inline docs/what_is_hip.rst]:
```suggestion
* Developers can specialize for the platform (CUDA or ROCm) to tune for
```

File: docs/what_is_hip.rst [hunk 12]
```diff
@@ -8,41 +8,85 @@
...
+HIP is designed to work seamlessly with the ROCm Runtime
+(:doc:`ROCr <ROCR-Runtime:index>`). HIP provides two components: those that run
+on the CPU, also known as host system, and those that run on GPUs, also referred
+to as device. The host-based code is used to create device buffers, move data
+between the host application and a device, launch the device code (also known
+as kernel), manage streams and events, and perform synchronization.
+The kernel language provides a way to develop massively parallel programs that
+run on GPUs, and provides access to GPU specific hardware capabilities.
+
+access GPU specific hardware
```
Comment (neon60) [inline docs/what_is_hip.rst]: @MKKnorr  there is a problem here.

File: docs/what_is_hip.rst (hunk 11 above)
Comment (MKKnorr) [reply docs/what_is_hip.rst]: Fixed

File: docs/what_is_hip.rst (hunk 12 above)
Comment (MKKnorr) [reply docs/what_is_hip.rst]: Fixed. Those were remains from refactoring :D

File: docs/what_is_hip.rst [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    :align: center
+
+* HIP is a thin API with little or no performance impact over coding directly
+  in NVIDIA CUDA or AMD :doc:`ROCm <rocm:what-is-rocm>`.
+* HIP enables coding in a single-source C++ programming language including
+  features such as templates, C++11 lambdas, classes, namespaces, and more.
+* Developers can specialize for the platform (CUDA or ROCm) to tune for
+  performance or handle tricky cases.
+
+:doc:`ROCm <rocm:what-is-rocm>` offers compilers (``clang``, ``hipcc``), code
```
Comment (randyh62) [inline docs/what_is_hip.rst]:
```suggestion
ROCm offers compilers (``clang``, ``hipcc``), code
```

**Review by randyh62 — APPROVED** [review]

**Review by anugodavar — APPROVED** [review]
Provided comments over chat and they are fixed.

**Pull request merged**